pennylane/optimize/spsa.py

```diff
@@ -0,0 +1,255 @@
+    >>> for n in range(max_iterations):
+    ...     params, energy = opt.step_and_cost(cost, params)
+
+    Example of hybrid classical-quantum workflow:
+
+    >>> dev = qml.device("default.qubit", wires=n_qubits)
+    >>> @qml.qnode(dev, interface=None)
+    >>> def layer_fn_spsa(inputs, weights):
+    ...     qml.AngleEmbedding(inputs, wires=range(n_qubits))
+    ...     qml.BasicEntanglerLayers(weights, wires=range(n_qubits))
+    ...     return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_qubits)]
+
+    >>> opt = qml.SPSAOptimizer(maxiter=max_iterations)
+    >>> tensor_in = tf.Variable([0.27507603, 0.3453423])
+    >>> params = tf.Variable([[3.97507603, 2.00854038],
+    ...                       [3.12950603, 3.00854038],
+    ...                       [1.17907603, 1.10854038],
+    ...                       [0.97507603, 1.00854038],
+    ...                       [1.25907603, 0.40854088]])
+    >>> tensor_out = tf.Variable([0,0])
+
+    >>> init = tf.compat.v1.global_variables_initializer()
+    >>> with tf.compat.v1.Session() as sess:
+    ...     sess.run(init)
+    ...     for _ in range(max_iterations):
+    ...         # Take step
+    ...         params_a, layer_res = opt.step_and_cost(layer_fn_spsa,
+    ...                                np.tensor(tensor_in.eval(sess), requires_grad=False),
+    ...                                np.tensor(params.eval(sess)))
+    ...         params.assign(params_a[1], sess)
+    ...         tensor_out.assign(layer_res, sess)
+
+
+
+
+    Keyword Args:
+        maxiter (int): the maximum number of iterations expected to be performed
+        alpha (float): An hyperparameter to calculate :math:`a_k=\frac{a}{(A+k+1)^\alpha}`
+            for each iteration. Its asymptotically optimal value is 1.0
+        gamma (float): An hyperparameter to calculate :math:`c_k=\frac{c}{(k+1)^\gamma}`
+            for each iteration. Its asymptotically optimal value is 1/6
+        c (float): An hyperparameter related to the expected noise. It should be
+            approximately the standard deviation of the expected noise on the cost function
+        A (float): The stability constant expected to be 10% of maximum number
+            of expected iterations
+        a (float): An hyperparameter expected to be small in noisy situations,
+            whose value could be :math:`\frac{mag(\Delta\theta)}{mag(g(\theta))}(A+1)^\alpha`
+    """
+    # pylint: disable-msg=too-many-arguments
+    def __init__(self, maxiter=200, alpha=0.602, gamma=0.101, c=0.2, A=None, a=None):
+        if not A:
+            self.A = maxiter * 0.1
+        if not a:
+            self.a = 0.05 * (self.A + 1) ** alpha
+        self.c = c
+        self.alpha = alpha
+        self.gamma = gamma
+        self.k = 0
+        self.ak = self.a / (self.A + 1 + 1.0) ** self.alpha
+
+    def step_and_cost(self, objective_fn, *args, **kwargs):
+        """Update the parameter array :math:`x` with one step of the optimizer and return
```

Comment:
Worth updating the $x$ notation in this docstring.

Reply:
Do you mean changing it to theta?

Reply:
Yes :)